core/solver/ir.cpp

```diff
@@ -160,21 +160,31 @@ std::unique_ptr<LinOp> Ir<ValueType>::conj_transpose() const
 
 template <typename ValueType>
 void Ir<ValueType>::apply_impl(const LinOp* b, LinOp* x) const
+{
+    this->apply_with_guess(b, x, this->get_apply_guess());
+}
+
+
+template <typename ValueType>
+void Ir<ValueType>::apply_with_guess(const LinOp* b, LinOp* x,
+                                     initial_guess guess) const
 {
     if (!this->get_system_matrix()) {
         return;
     }
+    this->modify_input(b, x, guess);
```

Comment:
Why is this necessary? Doesn't apply_dense_impl already take care of that?

Reply:
no, apply_dense_impl only takes the information to eliminate the spmv computation.
but we still need to prepare the x with the correct value.
I modify the value before advanced_apply because I would like to make it give the same behavior as the default implementation.

Reply:
That makes sense, thanks! I think in the long run, we should aim to eliminate the need for this, since we can fuse the initialization of `x` with other operations, but for now, that looks good.